core/module.go

```diff
@@ -99,110 +67,46 @@ func (*Module) TypeDescription() string {
-		panic(err)
-	}
-	return call.NewModule(mod.InstanceID, mod.Name(), ref, pin)
-}
-
-func (mod *Module) Initialize(ctx context.Context, oldID *call.ID, newID *call.ID, dag *dagql.Server) (*Module, error) {
-	modName := mod.Name()
-	newMod := mod.Clone()
-	newMod.InstanceID = oldID // updated to newID once the call to initialize is done
-
-	// construct a special function with no object or function name, which tells
-	// the SDK to return the module's definition (in terms of objects, fields and
-	// functions)
-	getModDefFn, err := newModFunction(
-		ctx,
-		newMod.Query,
-		newMod,
-		nil,
-		newMod.Runtime,
-		NewFunction("", &TypeDef{
-			Kind:     TypeDefKindObject,
-			AsObject: dagql.NonNull(NewObjectTypeDef("Module", "")),
-		}))
-	if err != nil {
-		return nil, fmt.Errorf("failed to create module definition function for module %q: %w", modName, err)
-	}
-
-	result, err := getModDefFn.Call(ctx, &CallOpts{Cache: true, SkipSelfSchema: true, Server: dag})
-	if err != nil {
-		return nil, fmt.Errorf("failed to call module %q to get functions: %w", modName, err)
-	}
-	inst, ok := result.(dagql.Instance[*Module])
-	if !ok {
-		return nil, fmt.Errorf("expected Module result, got %T", result)
-	}
-
-	newMod.InstanceID = newID
-	newMod.Description = inst.Self.Description
-	for _, obj := range inst.Self.ObjectDefs {
-		newMod, err = newMod.WithObject(ctx, obj)
+	var ref, pin string
+	switch mod.Source.Self.Kind {
+	case ModuleSourceKindLocal:
+		ref = filepath.Join(mod.Source.Self.Local.ContextDirectoryPath, mod.Source.Self.SourceRootSubpath)
+
+	case ModuleSourceKindGit:
+		ref = mod.Source.Self.Git.CloneRef
+		if mod.Source.Self.SourceRootSubpath != "" {
+			ref += "/" + strings.TrimPrefix(mod.Source.Self.SourceRootSubpath, "/")
+		}
+		if mod.Source.Self.Git.Version != "" {
+			ref += "@" + mod.Source.Self.Git.Version
+		}
+		pin = mod.Source.Self.Git.Commit
+
+	case ModuleSourceKindDir:
+		// FIXME: this is better than nothing, but no other code handles refs that
+		// are an encoded ID right now
+		var err error
+		ref, err = mod.Source.Self.ContextDirectory.ID().Encode()
 		if err != nil {
-			return nil, fmt.Errorf("failed to add object to module %q: %w", modName, err)
-		}
-	}
-	for _, iface := range inst.Self.InterfaceDefs {
-		newMod, err = newMod.WithInterface(ctx, iface)
-		if err != nil {
-			return nil, fmt.Errorf("failed to add interface to module %q: %w", modName, err)
-		}
-	}
-	for _, enum := range inst.Self.EnumDefs {
-		newMod, err = newMod.WithEnum(ctx, enum)
-		if err != nil {
-			return nil, fmt.Errorf("failed to add enum to module %q: %w", mod.Name(), err)
+			panic(fmt.Sprintf("failed to encode context directory ID: %v", err))
 		}
+
+	default:
+		panic(fmt.Sprintf("unexpected module source kind %q", mod.Source.Self.Kind))
 	}
-	newMod.InstanceID = newID
 
-	return newMod, nil
+	return call.NewModule(mod.InstanceID, mod.Name(), ref, pin)
+}
+
+func (mod *Module) Evaluate(context.Context) (*buildkit.Result, error) {
```

Comment:
I might be missing something, but where is `Sync` on a `Module`/`ModuleSource` used?

Edit: I see it used in https://github.com/dagger/dagger.io/pull/4242, might be worth making a note that it's used in daggerverse so we don't accidentally remove it.

(hopefully most of those cases are now gone, but that's one of things that used to make this hard to work with, since you had to check daggerverse to see if a field was being used or not)

Reply:
Sure yeah I'll add a comment saying they are important